Display wsgi-status and name/version of used server

pbauer opened the pull request.

mister-roboto commented:
@pbauer thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

mister-roboto commented:
@pbauer thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

pbauer commented:
@tschorr do you know how to find the name of the wsgi-server we use while the instance is running?

pbauer commented:
@jenkins-plone-org please run jobs

tschorr commented:
> @tschorr do you know how to find the name of the wsgi-server we use while the instance is running?

No, and I can't find anything about this in https://www.python.org/dev/peps/pep-3333/#id23. So I have some doubts that it is possible to do this in an explicit, straightforward way for all available middleware.

pbauer commented:
I changed the code to do some guessing using the request. At least it works fine for ZServer and waitress.

jensens commented:
Look good enough to me. Later improvement is always possible.

ale-rt commented:
I would suggest something like this to have transaltions:
```
    <p tal:define="
         server_info view/server_info;
         has_wsgi server_info/wsgi;
       ">
      <ul>
        <li>
          <tal:i18n i18n:translate="">WSGI:</tal:i18n>
          <span i18n:translate="" tal:condition="has_wsgi">On</span>
          <span i18n:translate="" tal:condition="not:has_wsgi">Off</span>
        </li>
        <li>
          <tal:i18n i19n:translate="">Server:</tal:i18n>
          <span>${server_info/server_name}</span>
          <span>${server_info/version}</span>
        </li>
      </ul>
    </p>
```
Use of i18n:name is a plus.

tschorr commented:
@jensens @pbauer why not, it's definitely worth a try. And it should also work for the ZServer case, so maybe the code could be shortened a bit.

pbauer commented:
> > We should also display if it is running on wsgi or ZServer in /@@overview-controlpanel
> 
> My assumption when reading this in #2763 was that we want to display we are running on WSGI full stop.

Yes, it should be the default but my thinking was that you can to opt-out by setting `wsgi=off` when running Python 2. That would give you ZServer and support for FTP and WebDAV.

tschorr commented:
> > > We should also display if it is running on wsgi or ZServer in /@@overview-controlpanel
> > 
> > 
> > My assumption when reading this in #2763 was that we want to display we are running on WSGI full stop.
> 
> Yes, it should be the default but my thinking was that you can to opt-out by setting `wsgi=off` when running Python 2. That would give you ZServer and support for FTP and WebDAV.

That's fine for me. I was just assuming we would only display either ZServer or WSGI and not bother about the actual WSGI middleware/server software used.

tschorr commented:
just tried uWSGI - It shows server: unknown for both uwsgi and http socket.

tschorr commented:
... gunicorn sets the SERVER_SOFTWARE variable so it will already contain version info:

gunicorn/__init__.py:8:SERVER_SOFTWARE = "gunicorn/%s" % __version__

jensens commented:
WSGI 2 proposal makes SERVER_SOFTWARE required: https://wsgi.readthedocs.io/en/latest/proposals-2.0.html#making-some-keys-required
I have no idea about the state of the proposal.

tschorr commented:
> I have no idea about the state of the proposal.

@jensens the note at the top of https://wsgi.readthedocs.io/en/latest/proposals-2.0.html#proposals-related-to-wsgi-2-0 doesn't sound very encouraging to me :-)

tschorr commented:
... but still I think using SERVER_SOFTWARE to get some information about the middleware/server software is a good idea and possibly the best option we have.

jensens commented:
> doesn't sound very encouraging to me

That is why I wrote: no idea.

pbauer commented:
Ok, I added i18n and changed it a little so that server_version is only shown when server_name is a package. In some cases (like gunicorn) server_name also holds the version.

pbauer commented:
@jenkins-plone-org please run jobs

tschorr commented on Products/CMFPlone/controlpanel/browser/overview.py:
```diff
@@ -45,6 +48,31 @@ def core_versions(self):
+
+        # check for ZServer
+        servers = getattr(getConfiguration(), 'servers', None)
+        if servers and 'ZServer' in servers[0].__module__:
```
ZServer also sets SERVER_SOFTWARE , so it could be handled in the same way as the WSGI servers. This would simplify the code.

pbauer replied:
Oh, I missed that. I update the code.

pbauer commented:
@jenkins-plone-org please run jobs
